### PR TITLE
bot: allow more fine-grained reply control

### PR DESCRIPTION
**Pull request by euank**
Only light manual testing done. I need to add more info in the last commit message too.

### CODE REVIEW

File: bot.js [hunk 1]
```diff
@@ -335,36 +343,82 @@ bot.isChannel = function(name) {
...
-  return function(args) {
-    var repStr = bot.stringifyArgs.apply(this, arguments);
+  // custom reply takes options of the following:
+  // {
+  //   lines: 2, // number of lines to spam
+  //   replaceNewLines: false, // whether to replace newlines with | while figuring out if it'll spam
+  //   pmExtra: false, // whether to PM the whole message if it overflows
+  // }
+  var customReply = function(opts, args) {
+    var maxLineChars = 460; // convenient lie; the true number is complicated due to having to account for nick-length etc
```
Comment (LinuxMercedes) [inline bot.js]: You might be able to get this from `node-irc`. See https://github.com/martynsmith/node-irc/blob/master/lib/irc.js#L1069

File: bot.js [hunk 2]
```diff
@@ -335,36 +343,82 @@ bot.isChannel = function(name) {
...
-    } else {
-      bot.callModuleFn('pmnotice', [bot.client.nick, to, repStr]);
+    args = Array.prototype.slice.call(arguments, 1);
+
+    // If it's a pm, all options get ignored
+    if(isPm) {
+      spamReply.apply(this, args);
+    }
+
+    var repStr = bot.stringifyArgs.apply(this, args);
```
Comment (LinuxMercedes) [inline bot.js]: Nit: I'd add a comment here saying something to the effect of "calculate number of messages to send these lines"?

File: bot.js [hunk 3]
```diff
@@ -335,36 +343,82 @@ bot.isChannel = function(name) {
...
-      bot.callModuleFn('pmnotice', [bot.client.nick, to, repStr]);
+  var reply = function(args) {
+    if(isPm) {
+      spamReply.apply(this, arguments);
+      return;
     }
-    bot.client.action(to, repStr);
+    // Default to not spam if it's not a pm
+    args = Array.prototype.slice.call(arguments);
+    args.unshift({lines: 2, replaceNewLines: false, pmExtra: false});
```
Comment (LinuxMercedes) [inline bot.js]: Is 2 enough lines here for stuff like wolfram alpha?

File: bot.js (hunk 1 above)
Comment (euank) [inline bot.js]: That's a convenient lie too I think, but I'll grab that value and call it good enough

File: bot.js (hunk 3 above)
Comment (euank) [inline bot.js]: I'll update wolfram and twitter to use `custom`

**Comment by euank** [comment]
Added some tests and reworked some of the logic based on writing the tests. Still not totally sure if this is all what I want... I'll write some more tests and then convert the modules that should use this over

**Comment by drusepth** [comment]
@euank What is the intended functionality here? Is it just some kind of throttling for how often it spits responses out to a channel?

**Comment by euank** [comment]
@drusepth no throttling, that's a configuration the irc-client lib can already do separately.

This is integrating the t_t anti-spam mechanism (akin to [this](https://github.com/euank/reckless-irc-exec/blob/43f60e75cff11f889c8fca83421de507e8cb5bec/bot.js#L15-L42)) in preparation for merging in similarly spammy code.

The intent is that modules, by default, when they call `reply(someString)` will have that string truncated to a reasonable length unless the module explicitly intends to spew more, either via multiple `reply` calls, or via `reply.spam` or `reply.custom`.

A good example of a module this helps with now is the `!ud` module. Because that's replying with the result of network traffic based on the user query, right now it's easy to spam a giant wall of text.

This will make that impossible unless the module explicitly opts into being able to spam so much.

File: bot.js [hunk 4]
```diff
@@ -335,36 +343,111 @@ bot.isChannel = function(name) {
...
-      bot.callModuleFn('channotice', [bot.client.nick, to, repStr]);
+      // This isn't fitting, say a stripped version
+      bot.client.say(to, withoutNewlines.substring(0, maxLineChars - 4) + " ...");
+      // don't return, we might have to spit out extra
     } else {
-      bot.callModuleFn('pmnotice', [bot.client.nick, to, repStr]);
+      // we know it doesn't fit like this already, let's say a trimmed down version.
+      let linesToSay = [];
+      let maxLength = maxLineChars * opts.lines - 4; // -4 to leave room for '...'
+      for(let i=0; i < lines.length; i++) {
```
Comment (LinuxMercedes) [inline bot.js]: I feel like this loop could be more clearly done as a `reduce`, but maybe that's just the haskell in my brain talking.

Comment (euank) [inline bot.js]: Because of the `i == (opts.lines - 1)` bit it seemed a little less clearcut.

Comment (LinuxMercedes) [inline bot.js]: Ahh, I see. Maybe a `map()` +`take(opts.lines)`?

(Really, though, this is just a nice bikeshed. If you're happy with it, I'm happy with it =D)

**Pull request merged**